lib/geminabox/dependency_info.rb

```diff
@@ -0,0 +1,161 @@
+# frozen_string_literal: true
+
+module Geminabox
+  class DependencyInfo
+    attr_reader :gem_name
+
+    def initialize(gem_name)
+      @gem_name = gem_name
+      @versions = nil
+      @content = nil
+      yield self if block_given?
+    end
+
+    def versions
+      @versions ||= decode
+    end
+
+    def version_names
+      versions.map { |version, platform, _, _| version_name(version, platform) }
+    end
+
+    def versions=(versions)
+      @versions ||= versions
+      @content = nil
+    end
+
+    def content
+      @content ||= encode
+    end
+
+    def content=(content)
+      @content = content
+      @versions = nil
+    end
+
+    def digest
+      Digest::MD5.hexdigest(content)
+    end
+
+    def checksums
+      versions.map { |v| checksum(v) }
+    end
+
+    def add_gem_spec_and_gem_checksum(spec, checksum)
+      raise ArgumentError, "can't add spec for gem #{spec.name} to gem info for gem #{gem_name}" if spec.name != gem_name
+
+      platform = spec.platform.to_s if spec.platform && spec.platform != 'ruby'
+      dependencies = spec.runtime_dependencies.sort.map { |dep| [dep.name, dep.requirement.requirements.sort.map { |a| a.join(" ") }] }
+      requirements = [['checksum', [checksum]]]
+      requirements += constraints_for('ruby', spec.required_ruby_version)
+      requirements += constraints_for('rubygems', spec.required_rubygems_version)
+      add_gem_version(spec.version.to_s, platform, dependencies, requirements)
+      @content = nil
+    end
+
+    def remove_gem_spec(spec)
+      raise ArgumentError, "can't remove spec for gem #{spec.name} from gem info for gem #{gem_name}" if spec.name != gem_name
+
+      version_to_delete = GemVersion.from_spec(spec)
+
+      versions.reject! do |version, platform, _, _|
+        version_to_delete == gem_version(version, platform)
+      end
+      @content = nil
+    end
+
+    def subsumed_by?(other)
+      Set.new(checksums) <= Set.new(other.checksums)
+    end
+
+    def disjoint?(other)
+      !Set.new(checksums).intersect?(Set.new(other.checksums))
+    end
+
+    def conflicts(other)
+      other_checksums = Set.new(other.checksums)
+      local_versions = versions.reject { |v| other_checksums.include?(checksum(v)) }
+      local_versions.map { |v, p, _, _| version_name(v, p) }
+    end
+
+    private
+
+    PREAMBLE = "---\n"
+
+    def decode(content = @content)
+      content.to_s.each_line.map { |line| parse_line(line.chomp) unless line == PREAMBLE }.compact
+    end
+
+    def encode(versions = @versions)
+      str = PREAMBLE.dup
```

Comment:
since ruby 2.2 support has been dropped, we can use
```suggestion
      str = +PREAMBLE
```

Reply:
Rubocop does not like it, that's why I changed it.

Reply:
Also, I really did not know this pattern before I saw it in your code.

Reply:
ok, I see, it won't make a difference here anyway...